pkg/receive/handler.go

```diff
@@ -434,7 +437,47 @@ func (h *Handler) parallelizeRequests(ctx context.Context, tenant string, replic
 		close(ec)
 	}()
 
-	return n, ec
+	// At the end, make sure to exhaust the channel, letting remaining unnecessary requests finish asnychronously.
+	// This is needed if context is cancelled or if we reached success of fail quorum faster.
+	defer func() {
+		go func() {
+			for err := range ec {
+				if err != nil {
+					level.Debug(h.logger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
+				}
+			}
+		}()
+	}()
+
+	var (
+		success int
+		errs    terrors.MultiError
+	)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-ec:
+			if err == nil {
+				success++
+				if success >= successQuorum {
+					// Quorum met, claim success.
+					return nil
+				}
+				continue
+			}
+			errs.Add(err)
+			if countCause(errs, isNotReady) >= len(wreqs)-successQuorum {
```

Comment:
that calculation on the right side is not the same, as it was before, quorum is quorum, whether it's success or failure